compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2466,6 +2466,19 @@ impl<'a> Parser<'a> {
         if !self.eat_keyword(kw::In) {
             self.error_missing_in_for_loop();
         }
+
+        // Recover from missing expression in `for` loop
+        if self.token == token::OpenDelim(token::Delimiter::Brace)
+            && self.prev_token.is_keyword(kw::In)
+            && self.may_recover()
+        {
+            let mut err = self.struct_span_err(self.token.span, "expected expression, found block");
+            err.help(
+                "try adding an expression to the `for` loop: `for PATTERN in EXPRESSION { BODY }`",
+            );
+            return Err(err);
+        }
```

Comment:
Doesn't this recovery break the following example:

```rust
for _ in { 0..10 } {}
```

Reply:
Will break it, How can i avoid that?

Reply:
Only emit an error if:
1. The `loop_block` that is being parsed at 2487 doesn't start with `{` (you can do the check on 2486)
2. *and* the `expr` we parsed on 2483 is a block expression.

Reply:
Thanks :)